File: apps/web-ui/lib/workspace.ts
export type LandingRole =
  | "Owner"
  | "Admin"
  | "PlantManager"
  | "Planner"
  | "Store"
  | "Dispatch"
  | "Sales"
  | "Operator"

type QuickAction = {
  href: string
  label: string
  detail: string
}

export const LANDING_LABELS: Record<LandingRole, string> = {
  Owner: "Owner",
  Admin: "Admin",
  PlantManager: "Plant Manager",
  Planner: "Planner",
  Store: "Store",
  Dispatch: "Dispatch",
  Sales: "Sales",
  Operator: "Operator",
}

export const ROLE_PRIORITY: LandingRole[] = ["Owner", "Admin", "PlantManager", "Planner", "Store", "Dispatch", "Sales", "Operator"]

// Legacy role aliases are normalized here only so old auth/session rows land on a canonical workspace.
// User-facing navigation policy should use the condensed role matrix above.
export const ROLE_TO_LANDING: Record<string, LandingRole> = {
  Owner: "Owner",
  Admin: "Admin",
  PlantManager: "PlantManager",
  SupervisorEntry: "PlantManager",
  Production: "PlantManager",
  QC: "PlantManager",
  Planner: "Planner",
  Store: "Store",
  Dispatch: "Dispatch",
  DispatchMaker: "Dispatch",
  DispatchApprover: "Dispatch",
  Sales: "Sales",
  SOMaker: "Sales",
  SOApprover: "Sales",
  Operator: "Operator",
}

export function resolveLandingRole(roles: string[] = []): LandingRole {
  const mapped = new Set(roles.map((role) => ROLE_TO_LANDING[role]).filter(Boolean))
  return ROLE_PRIORITY.find((role) => mapped.has(role)) || "Operator"
}

export function rolesForSwitcher(roles: string[] = []) {
  const unique = Array.from(new Set(roles.map((role) => ROLE_TO_LANDING[role] || role).filter(Boolean)))
  return ROLE_PRIORITY.filter((role) => unique.includes(role))
}

export function landingPathForRole(role: string | null | undefined) {
  const landingRole = ROLE_TO_LANDING[String(role || "")] || resolveLandingRole(role ? [role] : [])
  if (landingRole === "Owner") return "/landing/owner"
  if (landingRole === "Admin") return "/landing/admin"
  if (landingRole === "Planner") return "/planning/board/winder"
  if (landingRole === "PlantManager") return "/planning/tracker"
  if (landingRole === "Sales") return "/sales-orders"
  if (landingRole === "Store") return "/inventory"
  if (landingRole === "Dispatch") return "/logistics/dispatch"
  return "/dashboard"
}

export const LANDING_QUICK_ACTIONS: Record<LandingRole, QuickAction[]> = {
  Owner: [
    { href: "/reports/owner", label: "Owner Pack", detail: "Review KPI stack, OTIF, and exceptions." },
    { href: "/planning", label: "Planning Board", detail: "Check route loading and bottlenecks." },
    { href: "/production/job-cards", label: "Job Cards", detail: "Validate active cards and stage completion." },
    { href: "/production/reconciliation", label: "Reconciliation", detail: "Close cost and material variances." },
  ],
  Admin: [
    { href: "/system/users", label: "Role Matrix", detail: "Manage users, roles, and governance." },
    { href: "/quality", label: "Quality Lifecycle", detail: "Review holds, inspections, and release decisions." },
    { href: "/reports/owner", label: "Owner Pack", detail: "Monitor cross-plant operations." },
    { href: "/planning", label: "Planning Board", detail: "Inspect route-level readiness." },
  ],
  PlantManager: [
    { href: "/planning/board?section=winder", label: "Winder Plan", detail: "Schedule machine and shift assignments." },
    { href: "/quality", label: "Quality Desk", detail: "Log inspections and release active holds." },
    { href: "/production/job-cards", label: "Job Cards", detail: "Track cards through the production spine." },
    { href: "/dispatch", label: "Dispatch Ready", detail: "Confirm finished jobs ready for handoff." },
  ],
  Planner: [
    { href: "/sales-orders", label: "Sales Queue", detail: "Review approvals and pending releases." },
    { href: "/planning", label: "Planning Workspace", detail: "Plan by stage, machine, and shift." },
    { href: "/planning/tracker", label: "Tracker", detail: "Monitor WIP and delays by stage." },
    { href: "/specifications/new", label: "Spec Sheet", detail: "Create recipe-backed specification sheets." },
  ],
  Store: [
    { href: "/inventory", label: "Inventory Actions", detail: "Run inward, issue, and reservations." },
    { href: "/inventory/genealogy", label: "Genealogy", detail: "Trace reel lineage, issues, and scan events." },
    { href: "/inventory/valuation", label: "Valuation", detail: "Audit stock value and risk items." },
    { href: "/inventory/reels/inward", label: "Reel Inward", detail: "Record reel-wise inward entries." },
  ],
  Dispatch: [
    { href: "/logistics/dispatch", label: "Dispatch Desk", detail: "Move finished-goods lots into challans." },
    { href: "/inventory/reservations", label: "FG Reservations", detail: "Check stock reserved against open customer demand." },
    { href: "/planning/tracker", label: "Ready Tracker", detail: "See jobs moving toward dispatch readiness." },
    { href: "/reports/dispatch", label: "Dispatch Reports", detail: "Review shipped quantity and pending handoff." },
  ],
  Sales: [
    { href: "/sales-orders/new", label: "Create Sales Order", detail: "Capture PO demand and release needs." },
    { href: "/sales-orders", label: "Sales Orders", detail: "Approve, release, and track line items." },
    { href: "/reports/sales", label: "Sales Reports", detail: "Track OTIF and delayed commitments." },
    { href: "/dispatch", label: "Dispatch Status", detail: "Check commercial handoff status." },
  ],
  Operator: [
    { href: "/production/supervisor-entry", label: "QR / Stage Entry", detail: "Scan job card and enter stage output." },
    { href: "/production/job-cards", label: "Assigned Job Cards", detail: "Open printable job-card details." },
    { href: "/planning/tracker", label: "Stage Tracker", detail: "Check where scanned jobs are standing." },
    { href: "/dashboard", label: "My Work", detail: "Return to the operator landing." },
  ],
}

File: hariom-erp/services/auth-service/src/workspace.py
from __future__ import annotations

from typing import Iterable

LANDING_PRIORITY = [
    "Owner",
    "Admin",
    "PlantManager",
    "Planner",
    "Store",
    "Dispatch",
    "Sales",
    "Operator",
]

BUSINESS_ROLE_ORDER = LANDING_PRIORITY.copy()
BUSINESS_ROLE_SET = set(BUSINESS_ROLE_ORDER)

ROLE_TO_LANDING = {
    "Owner": "Owner",
    "Admin": "Admin",
    "PlantManager": "PlantManager",
    "Planner": "Planner",
    "SpecMaker": "Planner",
    "SpecApprover": "Planner",
    "Production": "PlantManager",
    "Operator": "Operator",
    "SupervisorEntry": "PlantManager",
    "Store": "Store",
    "Dispatch": "Dispatch",
    "DispatchMaker": "Dispatch",
    "DispatchApprover": "Dispatch",
    "Sales": "Sales",
    "SOMaker": "Sales",
    "SOApprover": "Sales",
    "QC": "PlantManager",
}

LEGACY_ROLE_NAMES = sorted({role for role in ROLE_TO_LANDING if role not in BUSINESS_ROLE_SET})

LANDING_LABELS = {
    "Owner": "Owner",
    "Admin": "Admin",
    "PlantManager": "Plant Manager",
    "Planner": "Planner",
    "Store": "Store",
    "Dispatch": "Dispatch",
    "Sales": "Sales",
    "Operator": "Operator",
}

ROLE_CAPABILITIES = {
    "Owner": {
        "summary": "Full company visibility, dashboards, approvals, specs, stock close, and system oversight.",
        "permissions": [
            "analytics:view",
            "reports:view",
            "spec:create",
            "so:create",
            "so:approve",
            "planning:schedule",
            "production:entry",
            "production:close",
            "inventory:inward",
            "inventory:reserve",
            "inventory:close",
            "dispatch:create",
            "dispatch:approve",
            "dispatch:validate",
            "master:manage",
            "supplier:manage",
            "location:manage",
            "system:manage",
        ],
    },
    "Admin": {
        "summary": "System setup, users, plants, machines, locations, masters, and role governance.",
        "permissions": [
            "analytics:view",
            "reports:view",
            "spec:create",
            "so:create",
            "so:approve",
            "planning:schedule",
            "production:entry",
            "production:close",
            "inventory:inward",
            "inventory:reserve",
            "inventory:close",
            "dispatch:create",
            "dispatch:approve",
            "dispatch:validate",
            "master:manage",
            "supplier:manage",
            "location:manage",
            "system:manage",
        ],
    },
    "Sales": {
        "summary": "Customer PO, sales orders, release requests, and commercial tracking.",
        "permissions": ["so:create", "so:approve", "reports:view"],
    },
    "Planner": {
        "summary": "Release queue, three-day machine planning, tracker, MRP signals, and schedule reports.",
        "permissions": ["planning:schedule", "so:approve", "analytics:view", "reports:view"],
    },
    "PlantManager": {
        "summary": "Supervisor entry, machine execution, stage completion, QC holds, and reconciliation.",
        "permissions": ["production:entry", "production:close", "inventory:reserve", "analytics:view", "reports:view"],
    },
    "Store": {
        "summary": "RM inward, reel issue, locations, stock risk, opening load, and stock close support.",
        "permissions": ["inventory:inward", "inventory:reserve", "inventory:close", "supplier:manage", "reports:view"],
    },
    "Dispatch": {
        "summary": "Finished-goods readiness, challans, dispatch validation, and customer handoff.",
        "permissions": ["dispatch:create", "dispatch:approve", "dispatch:validate", "reports:view"],
    },
    "Operator": {
        "summary": "QR scan, assigned job card, and simple floor input only.",
        "permissions": ["operator:scan", "production:entry"],
    },
}

OVERRIDE_RIGHTS = [
    {"key": "sales", "label": "Sales order rights", "roles": ["Sales"]},
    {"key": "planner", "label": "Planner board rights", "roles": ["Planner"]},
    {"key": "plant_floor", "label": "Supervisor / plant floor rights", "roles": ["PlantManager"]},
    {"key": "store", "label": "Inventory and stock-close rights", "roles": ["Store"]},
    {"key": "dispatch", "label": "Dispatch rights", "roles": ["Dispatch"]},
    {"key": "operator", "label": "QR operator rights", "roles": ["Operator"]},
    {"key": "reports", "label": "Reports and analytics visibility", "roles": ["Owner"]},
    {"key": "system", "label": "System setup rights", "roles": ["Admin"]},
]


def canonical_role_name(role: str | None) -> str | None:
    if not role:
        return None
    text = str(role).strip()
    return ROLE_TO_LANDING.get(text, text if text in BUSINESS_ROLE_SET else None)


def resolve_landing_role(role_names: Iterable[str]) -> str | None:
    normalized = {canonical_role_name(role) for role in role_names if canonical_role_name(role)}
    for landing in LANDING_PRIORITY:
        if landing in normalized:
            return landing
    return None


def landing_label(landing_role: str | None) -> str | None:
    if not landing_role:
        return None
    return LANDING_LABELS.get(landing_role, landing_role)


def seeded_role_groups() -> list[dict[str, object]]:
    grouped: list[dict[str, object]] = []
    for landing_role in LANDING_PRIORITY:
        grouped.append(
            {
                "landing_role": landing_role,
                "landing_label": landing_label(landing_role),
                "roles": [landing_role],
                "summary": ROLE_CAPABILITIES[landing_role]["summary"],
                "permissions": ROLE_CAPABILITIES[landing_role]["permissions"],
            }
        )
    return grouped
